src/rgw/rgw_main.cc

```diff
@@ -227,8 +227,23 @@ int radosgw_Main(int argc, const char **argv)
   // privileged ports
   flags |= CINIT_FLAG_DEFER_DROP_PRIVILEGES;
 
+  // Load the config from the files, but not the mon
+  global_pre_init(&defaults, args, CEPH_ENTITY_TYPE_CLIENT,
+		  CODE_ENVIRONMENT_DAEMON, flags);
+
+  // Get the store backend
+  const auto& config_store = g_conf().get_val<std::string>("rgw_backend_store");
+
+  cerr << "config_store: " << config_store << std::endl;
+  if ((config_store == "dbstore") ||
```

Comment:
this is fine for immediate purposes, but how would we handle this with dynamic stores?  would a command line option be appropriate here?

Reply:
(more than fine, I was expecting a lot more to be involved :)

Reply:
I was expecting a lot more, too.  I'm not sure what to do about dynamic stores.  I have a few ideas, though.
1. Require all stores to be defined at startup (not, not all filters, but all stores).  They need not be loaded, but defined.
2. Find some way to replace g_conf() with per-store config.  Then, each store can initialize it's config in the correct way.  This may need a lot of work in the config code, I don't know, and haven't looked yet.
3. Strict upgrade of g_conf().  That is, if a store is loaded that needs RADOS, tear down g_conf() and re-create it with the MON connection.
4. A config option to say whether this RGW should connect to a ceph cluster.  In general, either a ceph cluster is available, or it's not.  If it is, we can connect.  If it's not, we can't.  I'm fine not supporting adding a ceph cluster near an existing RGW without restarting that RGW.

Reply:
If #1 means we cannot define a new store at load-time, it seems like a non-starter.  #2 would be clean, if we could get away with it.  It seems like we know it probably mostly works.  Adam and others have certainly suggested to get rid of g_conf() usage in rgw.  Using the right handle in RGW contexts seems manageable?  #3 seems hacky.  #4 seems hacky, but easily fixable later?

Reply:
Why do we have to be able to load a completely unknown store at run-time?  Is there some reason we can't know at startup all available, supported stores?  Also, why do we need to support adding a ceph cluster to an already running RGW?  Is someone really going to start a stand-alone rgw, decide it's not good enough, install a full ceph cluster around it, then add a rados store to the running RGW, all without restarting it?

I'm trying to understand the usecases for stores.  My understanding of these usecases was much more limited than what the above comment implies.

(Note, none of this discussion applies to this PR, of course...)

Reply:
what I meant that I don't want, is to have to have a static check in the *core rgw* code that recognizes loadable drivers, I assume that's the plan

Reply:
Actually, I'll add this to the refactoring meeting agenda, rather than discussing it here.

Reply:
Okay, I just meant when the RGW is started, not at compile time.

Reply:
> 4\. A config option to say whether this RGW should connect to a ceph cluster.  In general, either a ceph cluster is available, or it's not.  If it is, we can connect.  If it's not, we can't.  I'm fine not supporting adding a ceph cluster near an existing RGW without restarting that RGW.

this is my favorite option. if there is a monitor available, all stores should have the option to take advantage of the mon's config database and its ability to push config changes at runtime

whatever is deploying rgw will have to know what type of store to use, and whether monitors are available. if there is no monitor, it'll have to generate a ceph.conf to pass in all of this config